### PR TITLE
Remove debug, add decay and merging of data only

### PR DESCRIPTION
**Pull request by mdubbyap**
- decay : the total digest count will converge to `bufferSize / (1 - decayFactor)`
  if we use `decayFactor` 0.9 and `bufferSize` 1000, this means total count 10000
  so 99th percentile will not be overly influenced by a few bad values and similarly
  the ranking/selection will not be (provided we use scale function which keeps small
  enough bins towards the top)

### CODE REVIEW

**Comment by cory-signalfx** [comment]
Why remove debug?

File: tdigest/tdigest.proto [hunk 1]
```diff
@@ -15,10 +15,10 @@ message MergingDigestData {
     double min = 3;
     double max = 4;
     double reciprocalSum = 5;
+    bool debug = 6;
 }
 
 message Centroid {
     double mean = 1;
     double weight = 2;
-    repeated double samples = 3;
```
Comment (cory-signalfx) [inline tdigest/tdigest.proto:23]: I'd have to brush up on my protobuf… can you safely remove stuff? :)

File: tdigest/merging_digest.go [hunk 2]
```diff
@@ -481,3 +482,32 @@ func (td *MergingDigest) Data() *MergingDigestData {
...
 	}
 }
+
+// decay decays the histo to make values at the top less interesting over time
+func (td *MergingDigest) decay(p float64) {
+	td.mergeAllTemps()
+	for _, c := range td.mainCentroids {
+		c.Weight = c.Weight * p
+		c.Mean = c.Mean * p
+	}
```
Comment (cory-signalfx) [inline tdigest/merging_digest.go:492]: This does what it says, but I have no idea if this is mathematically valid. :)

File: tdigest/merging_digest.go [hunk 3]
```diff
@@ -481,3 +482,32 @@ func (td *MergingDigest) Data() *MergingDigestData {
...
+}
+
+// Clone gets you a clone of this tdigest
+func (td *MergingDigest) Clone() *MergingDigest {
```
Comment (cory-signalfx) [inline tdigest/merging_digest.go:496]: No test?

File: tdigest/merging_digest.go [hunk 4]
```diff
@@ -388,6 +379,19 @@ func (td *MergingDigest) Merge(other *MergingDigest) {
 	td.reciprocalSum = oldReciprocalSum + other.reciprocalSum
 }
 
+// Merge another digest data into this one. Neither td nor other can be shared
+// concurrently during the execution of this method.
+func (td *MergingDigest) MergeData(other *MergingDigestData) {
```
Comment (cory-signalfx) [inline tdigest/merging_digest.go:384]: No uses of this in here, is it for something else? Tests?

File: tdigest/merging_digest.go [hunk 5]
```diff
@@ -36,8 +36,7 @@ type MergingDigest struct {
 	min           float64
 	max           float64
 	reciprocalSum float64
-
-	debug bool
```
Comment (cory-signalfx) [inline tdigest/merging_digest.go:40]: Why rip all this out instead of just not using it?

File: tdigest/merging_digest.go [hunk 6]
```diff
@@ -481,3 +482,32 @@ func (td *MergingDigest) Data() *MergingDigestData {
 		ReciprocalSum: td.reciprocalSum,
 	}
 }
+
+// decay decays the histo to make values at the top less interesting over time
+func (td *MergingDigest) decay(p float64) {
+	td.mergeAllTemps()
+	for _, c := range td.mainCentroids {
+		c.Weight = c.Weight * p
+		c.Mean = c.Mean * p
```
Comment (sstewart-signalfx) [inline tdigest/merging_digest.go:491]: I think we only want to decay the Weight. 
If we decay the Mean, then the digest would also shift towards smaller values, which doesn't seem correct

File: tdigest/tdigest.proto (hunk 1 above)
Comment (mdubbyap) [reply tdigest/tdigest.proto:23]: You can if you have no expectations on the two being compatible. This is a fork specifically for the smart gateway. I don't plan on submitting this back.

File: tdigest/merging_digest.go (hunk 2 above)
Comment (mdubbyap) [reply tdigest/merging_digest.go:492]: From the chief data scientist.

File: tdigest/merging_digest.go (hunk 3 above)
Comment (mdubbyap) [reply tdigest/merging_digest.go:496]: God call.

File: tdigest/merging_digest.go (hunk 4 above)
Comment (mdubbyap) [reply tdigest/merging_digest.go:384]: Needs a test. But yeah, just seemed terrible to instantiate a new merging digest just to merge if i had the data serialized.

File: tdigest/merging_digest.go (hunk 6 above)
Comment (mdubbyap) [reply tdigest/merging_digest.go:491]: Talk to joe. This was the decision.  We need a way to decay exceptional values over time.

File: tdigest/merging_digest.go (hunk 5 above)
Comment (mdubbyap) [reply tdigest/merging_digest.go:40]: It's a pretty small point, but int64 * every single centroid * ever single histogram in the smart gateway. It's 60-120mb per smart gateway. /shrug.

Comment (cory-signalfx) [reply tdigest/merging_digest.go:40]: Ah, great! (This would've made great fodder for the PR body. :smile:)

File: tdigest/merging_digest.go (hunk 2 above)
Comment (cory-signalfx) [reply tdigest/merging_digest.go:492]: Ack!

File: tdigest/tdigest.proto (hunk 1 above)
Comment (cory-signalfx) [reply tdigest/tdigest.proto:23]: Ah, ok! I didn't realize that. :)

Comment (cory-signalfx) [reply tdigest/tdigest.proto:23]: Unrelated, we could pluck this out and make a standalone lib. ;)

Comment (cory-signalfx) [reply tdigest/tdigest.proto:23]: (or they could)

Comment (mdubbyap) [reply tdigest/tdigest.proto:23]: Probably should even. This being one tiny portion of the veneur project.

File: tdigest/merging_digest.go (hunk 5 above)
Comment (mdubbyap) [reply tdigest/merging_digest.go:40]: Fair point.

**Comment by mdubbyap** [comment]
I'm going to close this, create a separate tdigest repo, and address the rest of the comments there. Thanks, i will add y'all to that one so we can continue.

**Pull request closed without merging**